code/modules/library/computers/base.dm

```diff
@@ -0,0 +1,93 @@
+#define MAX_BOOK_FLAGS	3	// maximum number of times a book can be flagged before being removed from results
+
+/obj/machinery/computer/library
+	name = "visitor computer"
+	anchored = 1
+	density = 1
+	icon_keyboard = ""
+	icon_screen = "computer_on"
+	var/screenstate = 0
+	var/page_num = 1
+	var/num_pages = 0
+	var/num_results = 0
+	var/datum/library_query/query = new()
+
+	icon = 'icons/obj/library.dmi'
+	icon_state = "computer"
+
+/obj/machinery/computer/library/proc/interact_check(var/mob/user)
+	if(stat & (BROKEN | NOPOWER))
+		return 1
+
+	if (!Adjacent(user))
+		if (!issilicon(user) && !isobserver(user))
+			user.unset_machine()
+			user << browse(null, "window=library")
+			return 1
+
+	user.set_machine(src)
+	return 0
+
+/obj/machinery/computer/library/proc/get_page(var/page_num)
+	var/searchquery = ""
+	var/where = 0
+	if(query)
+		if(query.title && query.title != "")
+			searchquery += " WHERE title LIKE '%[query.title]%'"
+			where = 1
+		if(query.author && query.author != "")
+			searchquery += " [!where ? "WHERE" : "AND"] author LIKE '%[query.author]%'"
+			where = 1
+		if(query.category && query.category != "")
+			searchquery += " [!where ? "WHERE" : "AND"] category LIKE '%[query.category]%'"
+			if(query.category == "Fiction")
+				searchquery += " AND category NOT LIKE '%Non-Fiction%'"
+			where = 1
```

Comment:
Due to this block of code, I'll regard all touchings of the `query` var as potentially dangerous